flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/DynamicRecordProcessor.java

```diff
@@ -0,0 +1,163 @@
+
+    generator.open(openContext);
+  }
+
+  @Override
+  public void processElement(T element, Context ctx, Collector<DynamicRecordInternal> out)
+      throws Exception {
+    this.context = ctx;
+    this.collector = out;
+    generator.convert(element, this);
+  }
+
+  @Override
+  public void collect(DynamicRecord data) {
+    boolean exists = tableCache.exists(data.tableIdentifier()).f0;
+    String foundBranch = exists ? tableCache.branch(data.tableIdentifier(), data.branch()) : null;
+
+    Tuple2<Schema, CompareSchemasVisitor.Result> foundSchema =
+        exists
+            ? tableCache.schema(data.tableIdentifier(), data.schema())
+            : TableMetadataCache.NOT_FOUND;
+
+    PartitionSpec foundSpec = exists ? tableCache.spec(data.tableIdentifier(), data.spec()) : null;
+
+    if (!exists
+        || foundBranch == null
+        || foundSpec == null
+        || foundSchema.f1 == CompareSchemasVisitor.Result.SCHEMA_UPDATE_NEEDED) {
+      if (immediateUpdate) {
+        Tuple3<Schema, CompareSchemasVisitor.Result, PartitionSpec> newData =
+            updater.update(data.tableIdentifier(), data.branch(), data.schema(), data.spec());
+        emit(collector, data, newData.f0, newData.f1, newData.f2);
+      } else {
+        int writerKey =
+            hashKeyGenerator.generateKey(
+                data,
+                foundSchema.f0 != null ? foundSchema.f0 : data.schema(),
+                foundSpec != null ? foundSpec : data.spec(),
+                data.rowData());
+        context.output(
+            updateStream,
+            new DynamicRecordInternal(
+                data.tableIdentifier().toString(),
+                data.branch(),
+                data.schema(),
+                data.rowData(),
+                data.spec(),
+                writerKey,
+                data.upsertMode(),
+                DynamicSinkUtil.getEqualityFieldIds(data.equalityFields(), data.schema())));
+      }
+    } else {
+      emit(collector, data, foundSchema.f0, foundSchema.f1, foundSpec);
+    }
+  }
+
+  private void emit(
+      Collector<DynamicRecordInternal> out,
+      DynamicRecord data,
+      Schema schema,
+      CompareSchemasVisitor.Result result,
+      PartitionSpec spec) {
+    RowData rowData =
+        result == CompareSchemasVisitor.Result.SAME
+            ? data.rowData()
+            : RowDataEvolver.convert(data.rowData(), data.schema(), schema);
+    int writerKey = hashKeyGenerator.generateKey(data, schema, spec, rowData);
+    String tableName = data.tableIdentifier().toString();
+    out.collect(
+        new DynamicRecordInternal(
+            tableName,
+            data.branch(),
+            schema,
+            rowData,
+            spec,
+            writerKey,
+            data.upsertMode(),
+            DynamicSinkUtil.getEqualityFieldIds(data.equalityFields(), schema)));
+  }
+
+  @Override
+  public void close() {}
```

Comment:
Is this intentional, that the `super.close()` method not called? I see that you call super methods in several place, but not here.

Reply:
`ProcessFunction` also has a `close` (which is currently empty)

Reply:
True. Changed in 522d4bb. Needed to try/catch because Collector#close() doesn't throw Exception.